lib/trino-filesystem-s3/src/test/java/io/trino/filesystem/s3/TestS3FileSystemAwsS3.java

```diff
@@ -107,4 +119,50 @@ void testS3FileIteratorFileEntryTags()
             }
         }
     }
+
+    @Test
+    public void testSseCEncryptionRoundTrip()
+            throws NoSuchAlgorithmException, IOException
+    {
+        KeyGenerator generator = KeyGenerator.getInstance("AES");
+        generator.init(256);
+
+        String customerKey = Base64.getEncoder().encodeToString(generator.generateKey().getEncoded());
+        Location encryptedLocation = createLocation("encrypted");
+
+        TrinoFileSystem encryptedFileSystem = createS3FileSystemFactory().create(ConnectorIdentity
+                .forUser("test")
+                .withExtraCredentials(Map.of(S3FileSystemConstants.EXTRA_CREDENTIALS_SSEC_KEY, customerKey))
+                .build());
+
+        encryptedFileSystem.newOutputFile(encryptedLocation).createOrOverwrite("encoded content".getBytes(UTF_8));
+
+        // GET without key
+        assertThatThrownBy(() -> readFully(getFileSystem().newInputFile(encryptedLocation).newStream()))
+                .isInstanceOf(IOException.class)
+                .hasStackTraceContaining("The object was stored using a form of Server Side Encryption")
+                .hasMessage("Failed to open S3 file: s3://%s/encrypted".formatted(bucket()));
+
+        // HEAD without key
+        assertThatThrownBy(() -> getFileSystem().newInputFile(encryptedLocation).exists())
+                .isInstanceOf(IOException.class)
+                .hasMessage("S3 HEAD request failed for file: s3://%s/encrypted".formatted(bucket()));
+
+        // GET with key
+        assertThat(new String(readFully(encryptedFileSystem.newInputFile(encryptedLocation).newStream()), UTF_8))
+                .isEqualTo("encoded content");
```

Comment:
It's easier to read if you do the comparison against the actual value. Also, can use the newer `InputStream.readAllBytes()`
```suggestion
        assertThat(encryptedFileSystem.newInputFile(encryptedLocation).newStream().readAllBytes())
                .isEqualTo("encoded content".getBytes(UTF_8));
```

Reply:
Or we could try the `assertThat(InputStream)` method:
```suggestion
        assertThat(encryptedFileSystem.newInputFile(encryptedLocation).newStream())
                .asString(UTF_8).isEqualTo("encoded content");
```